cep-20.1.md

```diff
@@ -0,0 +1,287 @@
+  e.g. `- "{{ version }}"` in order for it to be valid YAML
+- Selectors use a YAML dictionary style (vs. comments in conda-build). E.g. `-
+  sel(osx): somepkg` instead of `- somepkg  # [osx]`
+
+## Selectors
+
+Selectors in the new spec take the following format:
+
+`sel(unix): selected_value`
+
+This is a valid YAML dictionary. Selector contents are simple boolean
+expressions and follow Python syntax. The following selectors are all valid:
+
+```
+win and arm64
+(osx or linux) and aarch64
+something == "test"
+```
+
+### The cmp function for variant selection
+
+Furthermore, we have a special "cmp" function that can be used to run a check
+against a selected variant version. The `cmp` function looks like the following:
+
+```
+cmp(python, "3.6")
+cmp(python, ">=3.6")
+cmp(python, ">=3.8,<3.10")
+etc
+```
+
+This can be used in a selector like so:
+
+```
+requirements:
+  build:
+    - sel(cmp(python, ">=3.6,<3.10")): dataclasses
+```
+
+This functionality generalizes and replaces the previous special variables such
+as `py2k`, `py3k`, `py36`, `py37`, and works just as well for NumPy, Ruby, R, or
+any other variant that might be of interest in the future.
+
+### Preprocessing selectors
+
+You can add selectors to any item, and the selector is evaluated in a
+preprocessing stage. If a selector evaluates to `true`, the item is flattened
+into the parent element. If a selector evaluates to `false`, the item is
+removed.
+
+```yaml
+source:
+  - sel(not win):
+      url: http://path/to/unix/source
+  - sel(win):
+      url: http://path/to/windows/source
+```
+
+Because the selector is a valid Jinja expression, complicated logic is possible:
+
+```yaml
+source:
+  - sel(win):
+      url: http://path/to/windows/source
+  - sel(unix and cmp(python, "2")):
+      url: http://path/to/python2/unix/source
+  - sel(unix and cmp(python, "3")):
+      url: http://path/to/python3/unix/source
+```
+
+Lists are automatically "merged" upwards, so it is possible to group multiple
+items under a single selector:
+
+```yaml
+test:
+  commands:
+    - sel(unix):
+      - test -d ${PREFIX}/include/xtensor
+      - test -f ${PREFIX}/include/xtensor/xarray.hpp
+      - test -f ${PREFIX}/lib/cmake/xtensor/xtensorConfig.cmake
+      - test -f ${PREFIX}/lib/cmake/xtensor/xtensorConfigVersion.cmake
+    - sel(win):
+      - if not exist %LIBRARY_PREFIX%\include\xtensor\xarray.hpp (exit 1)
+      - if not exist %LIBRARY_PREFIX%\lib\cmake\xtensor\xtensorConfig.cmake (exit 1)
+      - if not exist %LIBRARY_PREFIX%\lib\cmake\xtensor\xtensorConfigVersion.cmake (exit 1)
+
+# On unix this is rendered to:
+test:
+  commands:
+    - test -d ${PREFIX}/include/xtensor
+    - test -f ${PREFIX}/include/xtensor/xarray.hpp
+    - test -f ${PREFIX}/lib/cmake/xtensor/xtensorConfig.cmake
+    - test -f ${PREFIX}/lib/cmake/xtensor/xtensorConfigVersion.cmake
+```
+
+## Templating with Jinja
+
+The spec supports simple Jinja templating in the `recipe.yaml` file.
```

Comment:
Honestly since we are doing a bit of a redesign here, think it is worth asking whether we still want Jinja to play a role here. In other words, if we were not already using Jinja today, how would we have solved these use cases? How do others approach YAML templating? Could we think of pure YAML solutions?

Reply:
I agree, what does jinja accomplish that we can't do with the other new features?

Reply:
Well other new features would also have to be implemented and I am not sure we would necessary end up with something simpler.
I think there is a future where the entire build tool looks a bit different (and the recipe as well) and wolfi-os / melange have a nice approach: https://github.com/wolfi-dev/os/blob/main/libidn2.yaml

But I am not sure how to proceed here - we could either go down that road _right now_ or build something that is more reasonable for people to port existing recipes to and then iterate in a next step on the recipe format (when we have the necessary infrastructure nicely implemented in rattler-build).